### PR TITLE
[wasm] Compile .bc->.o in parallel, before passing to the linker

### PR DESCRIPTION
**Pull request by radical**
Currently, we use `mono-aot-cross` to compile the managed assemblies
into bitcode files. Then we pass the `.bc` files to `emcc` for linking.
We can save some amount of work, if precompile `.bc` files to `.o`
files, and pass them for linking.

This would also allow us to rebuild the `.o` files only when needed,
thus improving incremental builds.
Also, emscripten has recently removed some amount of parallelization, so
this allows us to do some pre-linking work in parallel, and speed up the
builds a little bit.

This is implemented via a new `EmccCompile` task, that is meant *only*
for compiling, and not as a general purpose `emcc` invocation task.

Also, this bumps msbuild used for tasks compilation to `16.9.0`.

### CODE REVIEW

**Comment by radical** [comment]
@akoeplinger Does the msbuild bump to 16.10.0 look fine?

**Review by lewing — APPROVED** [review]

**Comment by radical** [comment]
The browser builds keep failing with:

```
##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
D:\workspace\_work\1\s\src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj : error NU1201: Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0) [D:\workspace\_work\1\s\Build.proj]
##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
```

.. but I don't see this on `main` at all, AFAICS. But with the current state of CI, it's difficult to get a clear picture.

It might be related to the msbuild bump in this PR.

**Comment by radical** [comment]
@ViktorHofer Any idea why a msbuild bump to 16.10.0 in this PR might be failing wasm/android/iOS/etc builds? That's the only non-wasm specific change here.

> Some of the builds(browser, android, iOS etc) keep failing with:
> 
> ```
> ##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
> D:\workspace\_work\1\s\src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj : error NU1201: Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0) [D:\workspace\_work\1\s\Build.proj]
> ##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
> ```

File: src/mono/wasm/build/WasmApp.Native.targets [hunk 1]
```diff
@@ -165,18 +166,14 @@
       <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == ''">-Oz</_EmccOptimizationFlagDefault>
 
       <EmccCompileOptimizationFlag Condition="'$(EmccCompileOptimizationFlag)' == ''">$(_EmccOptimizationFlagDefault)</EmccCompileOptimizationFlag>
-      <EmccLinkOptimizationFlag    Condition="'$(EmccLinkOptimizationFlag)' == ''"   >$(_EmccOptimizationFlagDefault)</EmccLinkOptimizationFlag>
+      <EmccLinkOptimizationFlag    Condition="'$(EmccLinkOptimizationFlag)' == ''"   >-O0</EmccLinkOptimizationFlag>
```
Comment (radekdoulik) [inline src/mono/wasm/build/WasmApp.Native.targets:169]: We might also try to use `-Wl,-O0` and `-Wl,-lto-O0` here? (context: https://github.com/emscripten-core/emscripten/issues/14485#issuecomment-865213083)

Comment (radical) [reply src/mono/wasm/build/WasmApp.Native.targets:169]: Sure, we can try all that in a follow up PR. Unless, one of those is definitely better than `-O0`, then we can add it here itself.

**Comment by radical** [comment]
cc @ericstj @ViktorHofer for the change disabling `Microsoft.NETCore.Platforms.Tests` for wasm.

**Comment by radical** [comment]
1 test failure for wasm:
```
[04:55:37] info: browser: Console websocket connected.
[04:55:37] info: Arguments: --run,WasmTestRunner.dll,System.IO.Compression.Tests.dll,-notrait,category=IgnoreForCI,-notrait,category=OuterLoop,-notrait,category=failing
[04:55:37] fail: [out of order message from the browser]: http://127.0.0.1:34229/runtime.js 234:26 Uncaught TypeError: Cannot set property 'loaded_cb' of null
[05:10:37] fail: Application has finished with exit code TIMED_OUT but 0 was expected
```

@Daniel-Genkin is already looking at fixing this.
I think we can merge this.

**Comment by lewing** [comment]
> @Daniel-Genkin is already looking at fixing this.
> I think we can merge this.

fix should be in https://github.com/dotnet/runtime/pull/54587

**Comment by ViktorHofer** [comment]
> @ViktorHofer Any idea why a msbuild bump to 16.10.0 in this PR might be failing wasm/android/iOS/etc builds? That's the only non-wasm specific change here.

That very much sounds like https://github.com/NuGet/Home/issues/10368. Usually you can fix this by adding the transitive dependencies, in this case in the Platforms.Tests.csproj file.

**Comment by radical** [comment]
The failures are all unrelated to this PR.

**Comment by ViktorHofer** [comment]
> The failures are all unrelated to this PR.

Usually we link to issues that track the failures according to https://github.com/dotnet/runtime/blob/main/docs/issues-pr-management.md#common-policies. Can you please do so for the issues that you were seeing in the last build?

**Comment by radical** [comment]
Failures:

`Libraries Test Run checked coreclr windows x86 Release` - https://github.com/dotnet/runtime/issues/54628
`Libraries Test Run release mono OSX x64 Debug` - https://github.com/dotnet/runtime/issues/54624
`Libraries Test Run release coreclr windows x64 Debug` - https://github.com/dotnet/runtime/issues/54627
`Libraries Test Run release coreclr Linux x64 Debug` - https://github.com/dotnet/runtime/issues/54623

And for `CoreCLR Pri0 Runtime Tests Run windows arm64 checked` - `profiler.gc`, and `JIT.Directed` failed but I'm not able to access the logs for them.

**Pull request merged**